content/en/docs/Responses/error.md

```diff
@@ -21,21 +24,25 @@ description: >
 {{< /tab >}}
 {{< /tabpane >}}
 
-| Field |  Type | Req. | OpenS. | Details |
-| --- | --- | --- | --- | --- |
-| `code` | `int` | **Yes** |     | The error code |
-| `message` | `string` | No |     | The optional error message |
+| Field     | Type     | Req.    | OpenS. | Details                                                                                       |
+| --------- | -------- | ------- | ------ | --------------------------------------------------------------------------------------------- |
+| `code`    | `int`    | **Yes** |        | The error code                                                                                |
+| `message` | `string` | No      |        | The optional error message                                                                    |
+| `helpUrl` | `string` | No      | Yes    | A URL (documentation, configuration, etc) which may provide additional context for the error) |
 
 The following error codes are defined:
 
-| Code | Description |
-| --- | --- |
-| 0   | A generic error. |
-| 10  | Required parameter is missing. |
-| 20  | Incompatible Subsonic REST protocol version. Client must upgrade. |
-| 30  | Incompatible Subsonic REST protocol version. Server must upgrade. |
-| 40  | Wrong username or password. |
-| 41  | Token authentication not supported for LDAP users. |
-| 50  | User is not authorized for the given operation. |
-| 60  | The trial period for the Subsonic server is over. Please upgrade to Subsonic Premium. Visit subsonic.org for details. |
-| 70  | The requested data was not found. |
+| Code | Description                                                                                                           |
+| ---- | --------------------------------------------------------------------------------------------------------------------- |
+| 0    | A generic error.                                                                                                      |
+| 10   | Required parameter is missing.                                                                                        |
+| 20   | Incompatible Subsonic REST protocol version. Client must upgrade.                                                     |
+| 30   | Incompatible Subsonic REST protocol version. Server must upgrade.                                                     |
+| 40   | Wrong username or password.                                                                                           |
+| 41   | Token authentication not supported for LDAP users.                                                                    |
```

Comment:
Ah, it would probably be more correct to either
1. Roll it in to 42 and just say "authentication mechanism not supported"
2. Add a specific error to say token auth not supported. 41 (no LDAP) is actually different (since you do need password-based auth).
Thoughts?

Reply:
We can't really change the error 41 since it's legacy and would then differ between servers.

Making error 42 more generic as your 1) may be better for the future if we add another auth system.
But both works.

Reply:
As I see it, we *already have* broadened the meaning of the error 41 as that's clearly stated in the current documentation at https://opensubsonic.netlify.app/docs/opensubsonic-changes/. So now the documentation contradicts itself since this table has the original narrower definition of not supporting token authentication with LDAP while that other page tells to use the code 41 whenever token authentication is unsupported for any reason.

Of course this is strictly speaking an API break but most likely it will not cause any severe issues, because regardless of the reason of not supporting token+salt, the correct next steps for the client are the same.

If, on the other hand, the server would return 42 in case it doesn't support token+salt, then the non-OS clients might not be able to handle it. IMO that would be a a more severe API break.

Reply:
Error 41 was explicited in OS because some servers advertised subsonic API version > 1.13 and did not Implement token despite being mandatory in the API surface.
And since they returned different error codes we normalized to ensure clients know without breaking the login against the error 41 that means no token support. And we did not change the text to avoid changing the legacy surface without being necessary.

The new error 42 is tied to the new auths so can never be returned for a non OS client unless the server does not support token and password.
And then a non OS client can not connect at all. Way better to have a different error code in all cases to avoid legacy client trying to connect via passwords for ever.

So in this case nothing change for error 41 at all, and changing the text might have impact if a client relies on the text and not the error code.
Rare case and client should not but it's legacy why generate risk for 0 purpose.

The risk is really small, but you really need to understand the way OS is built and legacy history amongst all servers.

Anyway the risk being so small, I won't fight.

Reply:
> Wait, do you read this table in the documentation so that the server should always include the exact text of the "Description" in the message element of the response? Because this is not how I read it. And not even the original Subsonic server follows such practice in all cases. Sure, on some of the cases the message happens to be the same as the description but I think this is more a coincident than a documented part of the API. I honestly think that this table exists to tell the developers, what is the meaning of each numeric code, and not to tell what the message part of the response will be.

This is legacy and reason why we should not touch, each server and clients have made choices often different leading to the need for OS.
So since it's legacy the general default rule should be do not touch unless we are forced. Since many servers do return that exact error message it can be assumed that some bad clients did rely on the message. It should not, those clients are probably not maintained and probably not working anymore but still there's no reason to break them for no gain.

> I kind of get the feeling that we are maybe not understanding each other here and talking about different things. So do you now mean that if
a) the server doesn't support token+salt but does support username+password, then it should return 41
b) the server doesn't support either token+salt or username+password, but only supports API key, then it should return 42?

This is what @kgarner7 wrote and is in line with OS way of things. (https://deploy-preview-99--opensubsonic.netlify.app/docs/api-reference/)

Servers must return error 41 if they do not support token-based authentication, 42 if they do not support any other authentication mechanism (password-based and/or API key-based), and 43 if multiple conflicting authentication parameters are passed in at the same time.

Hence confirming that there was no need to change the text of 41 and why we brought a new specific error.